### PR TITLE
Replaced std::random_shuffle with std::shuffle in tri

### LINKED ISSUE #24010: c++17 removed random_shuffle
`std::random_shuffle` is used here:
https://github.com/matplotlib/matplotlib/blob/main/src/tri/_tri.cpp#L1469

It has been removed from c++17:
https://en.cppreference.com/w/cpp/algorithm/random_shuffle

> The reason for removing std::random_shuffle in C++17 is that the iterator-only version usually depends on [std::rand](https://en.cppreference.com/w/cpp/numeric/random/rand), which is now also discussed for deprecation. ([std::rand](https://en.cppreference.com/w/cpp/numeric/random/rand) should be replaced with the classes of the [<random>](https://en.cppreference.com/w/cpp/header/random) header, as [std::rand](https://en.cppreference.com/w/cpp/numeric/random/rand) is considered harmful.) In addition, the iterator-only std::random_shuffle version usually depends on a global state. The std::shuffle's shuffle algorithm is the preferred replacement, as it uses a URBG as its 3rd parameter.

### PR DESCRIPTION
**Pull request by tybeller**
## PR Summary
Replaced std::random_shuffle with std::shuffle for compliance with C++17. Deleted RandomNumberGenerator class. Included random library and used 32 bit mersenne twister as uniform random bit generator for shuffle. Note: TrapezoidMapTriFinder, triangulation interpolation, and refinement may yield different results due to this change. Closes #24010

## PR Checklist

**Tests and Styling**
- [N/A ] Has pytest style unit tests (and `pytest` passes).
- [N/A ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by tacaswell** [comment]
It looks like `std::shuffle` came in in c++11 which we just switch to having as our floor. It looks like moving that forwards was a good choice as it made this simple (I assume with enough `#ifdef` we could support a wider range of c++ versions....).

This seems reasonable and at least our tests are not sensitive to the different random order.

**Comment by ianthomas23** [comment]
Thanks @tybeller. It is usual to mention the issue that the PR closes within the text (not just within the title) so that there is a link between the two and the issue automatically closes when the PR is merged.

Also, I said in the issue that it would need a release note but I realise now that this is classified as a behaviour change ("API that stays valid but will yield a different result") which therefore needs an API change note. The `README` file in the `next_api_changes` directory describes what is required here, see https://github.com/matplotlib/matplotlib/blob/main/doc/api/next_api_changes/README.rst and let me know if this is not clear enough and I will help out.

**Comment by tybeller** [comment]
Thank you, sorry for any issues as this is my first time. If I'm reading this right I need to add the issue within the pull request and add and API change note correct? I'll get working on that and please let me know if there is anything else I'm missing.

**Comment by tybeller** [comment]
One question, is there a way to edit these changes into this pull request or should I make a separate pull request with these fixes?

**Comment by rcomer** [comment]
@tybeller if you edit your original post here and put “closes #24010“, that will link the issue to the PR, see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

Once you make any any changes to you branch and push them to github, they will show in this pull request.  There is advice for updating your branch in response to the review here: https://matplotlib.org/devdocs/devel/coding_guide.html#summary-for-pull-request-authors

File: doc/api/next_api_changes/behavior/24062-tb.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Change in TrapezoidMapTriFinder, triangulation interpolation, and refinement output due to change in shuffle.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+With std::random_shuffle removed from C++17, TrapezouidMapTriFinder, triangular interpolation, and refinement now use std::shuffle with a new random generator leading to a different output.
```
Comment (ianthomas23) [inline doc/api/next_api_changes/behavior/24062-tb.rst:3]:
```suggestion
With std::random_shuffle removed from C++17, TrapezoidMapTriFinder, triangular grid interpolation and refinement now use std::shuffle with a new random generator which may give different results.
```

**Review by ianthomas23 — CHANGES_REQUESTED** [review]
The code changes are fine, just a slight tweak to the API change note and this will be good to merge.

File: doc/api/next_api_changes/behavior/24062-tb.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Change in TrapezoidMapTriFinder, triangulation interpolation, and refinement output due to change in shuffle.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (oscargus) [inline doc/api/next_api_changes/behavior/24062-tb.rst:2]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
would solve the failing doc build (if I got the length right...).

File: doc/api/next_api_changes/behavior/24062-tb.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+With std::random_shuffle removed from C++17, TrapezoidMapTriFinder, triangular grid interpolation and refinement now use std::shuffle with a new random generator which may give different results.
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/24062-tb.rst:3]: Different results how? Does a plot change? Is the order different? This is too vague and as a user I have no idea whether I need to do anything here. Also, please word-wrap lines at 80 characters.

**Comment by ianthomas23** [comment]
@tybeller The definition of `class RandomNumberGenerator` at the end of `src/tri/_tri.h` can be removed.

**Comment by ianthomas23** [comment]
Here is a possible improved api change note:
```
``TrapezoidMapTriFinder`` uses different random number generator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The random number generator used to determine the order of insertion of
triangle edges in ``TrapezoidMapTriFinder`` has changed. This can result in a
different triangle index being returned for a point that lies exactly on an
edge between two triangles. This can also affect triangulation interpolation
and refinement algorithms that use ``TrapezoidMapTriFinder``.
```

**Comment by ianthomas23** [comment]
@tybeller Don't forget the changes in `src/tri/_tri.h`.

File: src/tri/_tri.cpp [hunk 4]
```diff
@@ -2055,17 +2056,4 @@ TrapezoidMapTriFinder::Trapezoid::set_upper_right(Trapezoid* upper_right_)
...
-    : _m(21870), _a(1291), _c(4621), _seed(seed % _m)
-{}
-
-unsigned long
-RandomNumberGenerator::operator()(unsigned long max_value)
-{
-    _seed = (_seed*_a + _c) % _m;
-    return (_seed*max_value) / _m;
-}
+}
```
Comment (oscargus) [inline src/tri/_tri.cpp:2059]: Please end the file with an empty line. That will fix the pre-commit issue. (The pyside6 issue is fixed in #24158.)

File: src/tri/_tri.h [hunk 5]
```diff
@@ -790,29 +790,3 @@ class TrapezoidMapTriFinder
...
-
-    // Return random integer in the range 0 to max_value-1.
-    unsigned long operator()(unsigned long max_value);
-
-private:
-    const unsigned long _m, _a, _c;
-    unsigned long _seed;
-};
-
-#endif
```
Comment (ianthomas23) [inline src/tri/_tri.h:818]: This line needs to be kept.

File: src/tri/_tri.h [hunk 6]
```diff
@@ -790,29 +790,3 @@ class TrapezoidMapTriFinder
 
     Node* _tree;    // Root node of the trapezoid map search tree.  Owned.
 };
```
Comment (oscargus) [inline src/tri/_tri.h:792]:
```suggestion
};

#endif
```

File: src/tri/_tri.h (hunk 5 above)
Comment (oscargus) [reply src/tri/_tri.h:818]: I took the liberty to add the line back again.

**Review by ianthomas23 — APPROVED** [review]

**Comment by tacaswell** [comment]
Thank you for this work @tybeller and congratulations on your first merged Matplotlib PR 🎉 Hopefully we will hear from you again!

**Pull request merged**